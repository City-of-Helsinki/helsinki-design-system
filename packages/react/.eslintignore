**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
