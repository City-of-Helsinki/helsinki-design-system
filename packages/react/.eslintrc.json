{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
