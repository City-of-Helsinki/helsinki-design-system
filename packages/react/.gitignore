node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
