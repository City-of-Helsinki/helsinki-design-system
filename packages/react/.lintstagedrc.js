module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
