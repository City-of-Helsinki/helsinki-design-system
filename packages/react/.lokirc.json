{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
