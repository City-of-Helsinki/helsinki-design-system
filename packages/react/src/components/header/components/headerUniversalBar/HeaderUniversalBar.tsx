import React, { cloneElement, useEffect } from 'react';

import '../../../../styles/base.module.css';
import styles from './HeaderUniversalBar.module.scss';
import { HeaderLink } from '../headerLink/HeaderLink';
import { useHeaderContext, useSetHeaderContext } from '../../HeaderContext';
import classNames from '../../../../utils/classNames';
import { getChildElementsEvenIfContainersInbetween } from '../../../../utils/getChildren';
import { AllElementPropsWithoutRef } from '../../../../utils/elementTypings';

export type HeaderUniversalBarProps = React.PropsWithChildren<
  AllElementPropsWithoutRef<'div'> & {
    /**
     * Additional class names to apply.
     */
    className?: string;
    /**
     * Children are expected to be HeaderLink components or a container with HeaderLink components inside.
     */
    children?: React.ReactNode;
    /**
     * ID of the header element.
     */
    id?: string;
    /**
     * Hypertext reference of the primary link.
     */
    primaryLinkHref?: string;
    /**
     * Link text for the primary link.
     */
    primaryLinkText?: string;
    /**
     * ARIA role to describe the contents.
     */
    role?: string;
  }
>;

export const HeaderUniversalBar = ({
  className,
  children,
  primaryLinkHref,
  primaryLinkText,
  ...rest
}: HeaderUniversalBarProps) => {
  const { isSmallScreen } = useHeaderContext();
  const childElements = getChildElementsEvenIfContainersInbetween(children);
  const { setUniversalContent } = useSetHeaderContext();

  useEffect(() => {
    const universalContent = getChildElementsEvenIfContainersInbetween(children);
    setUniversalContent(universalContent);
  }, [children]);

  if (isSmallScreen) return null;

  return (
    <div className={styles.headerUniversalBarContainer}>
      <div {...rest} className={classNames(styles.headerUniversalBar, className)}>
        <ul className={styles.headerUniversalBarList}>
          <li className={styles.universalBarMainLinkContainer}>
            <HeaderLink href={primaryLinkHref} label={primaryLinkText} className={styles.universalBarLink} />
          </li>
          {childElements.map((child, index) => {
            if (React.isValidElement(child)) {
              return (
                // eslint-disable-next-line react/no-array-index-key
                <li key={`secondary-link-${index}`} className={styles.universalBarSecondaryLinkContainer}>
                  {cloneElement(child as React.ReactElement, {
                    className: classNames(child.props.className, styles.universalBarLink),
                  })}
                </li>
              );
            }
            return null;
          })}
        </ul>
      </div>
    </div>
  );
};
