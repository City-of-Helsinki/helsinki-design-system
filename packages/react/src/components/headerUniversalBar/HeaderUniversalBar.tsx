import React, { Children, cloneElement, useContext } from 'react';

// import core base styles
import 'hds-core';
import styles from './HeaderUniversalBar.module.scss';
import { NavigationLink } from '../navigationLink';
import { HeaderContext } from '../header/HeaderContext';
import classNames from '../../utils/classNames';
import { getChildElements } from '../../utils/getChildren';

export type HeaderUniversalBarProps = React.PropsWithChildren<{
  /**
   * aria-label for describing universal bar.
   */
  ariaLabel?: string;
  /**
   * Additional class names to apply.
   */
  className?: string;
  /**
   * Children are expected to be NavigationLink components or a container with NavigationLink components inside.
   */
  children?: React.ReactNode;
  /**
   * ID of the header element.
   */
  id?: string;
  /**
   * Hypertext reference of the primary link.
   */
  primaryLinkHref: string;
  /**
   * Link text for the primary link.
   */
  primaryLinkText: string;
}>;

export const HeaderUniversalBar = ({
  ariaLabel,
  children,
  id,
  primaryLinkHref,
  primaryLinkText,
}: HeaderUniversalBarProps) => {
  const { isSmallScreen } = useContext(HeaderContext);
  if (isSmallScreen) return null;
  const childElements = getChildElements(children);

  return (
    <nav role="navigation" aria-label={ariaLabel} id={id} className={styles.headerUniversalBar}>
      <ul className={styles.headerUniversalBarList}>
        <li className={styles.universalBarMainLinkContainer}>
          <NavigationLink href={primaryLinkHref} label={primaryLinkText} className={styles.universalBarLink} />
        </li>
        {Children.map(childElements, (child, index) => {
          if (React.isValidElement(child)) {
            return (
              // eslint-disable-next-line react/no-array-index-key
              <li key={`secondary-link-${index}`} className={styles.universalBarSecondaryLinkContainer}>
                {cloneElement(child, {
                  className: classNames(child.props.className, styles.universalBarLink),
                })}
              </li>
            );
          }
          return null;
        })}
      </ul>
    </nav>
  );
};
