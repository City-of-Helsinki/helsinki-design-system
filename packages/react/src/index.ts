export * from './icons';
export * from './components';
export * from './ssr';
