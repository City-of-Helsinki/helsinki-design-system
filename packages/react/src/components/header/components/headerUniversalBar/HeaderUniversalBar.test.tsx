import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { HeaderUniversalBar } from './HeaderUniversalBar';

describe('<HeaderUniversalBar /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<HeaderUniversalBar primaryLinkText="hel.fi" primaryLinkHref="#" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<HeaderUniversalBar primaryLinkText="hel.fi" primaryLinkHref="#" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
