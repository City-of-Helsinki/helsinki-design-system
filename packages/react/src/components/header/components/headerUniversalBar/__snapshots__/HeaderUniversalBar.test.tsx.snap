// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderUniversalBar /> spec renders the component 1`] = `
<DocumentFragment>
  <nav
    class="headerUniversalBar"
    role="navigation"
  >
    <ul
      class="headerUniversalBarList"
    >
      <li
        class="universalBarMainLinkContainer"
      >
        <span
          class="navigationLinkWrapper"
        >
          <a
            class="link linkM navigationLink universalBarLink"
            href="#"
          >
            hel.fi
          </a>
        </span>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
